melior/src/ir/operation.rs

```diff
@@ -171,6 +171,36 @@ impl<'c> Operation<'c> {
             .map(|index| self.successor(index).expect("valid successor index"))
     }
 
+    /// Gets the number of attributes.
+    pub fn attribute_count(&self) -> usize {
+        unsafe { mlirOperationGetNumAttributes(self.raw) as usize }
+    }
+
+    /// Gets a attribute at a position.
+    pub fn attribute_at(&self, index: usize) -> Result<(Identifier<'c>, Attribute<'c>), Error> {
+        if index < self.attribute_count() {
+            unsafe {
+                let named_attribute = mlirOperationGetAttribute(self.raw, index as isize);
+                Ok((
+                    Identifier::from_raw(named_attribute.name),
+                    Attribute::from_raw(named_attribute.attribute),
+                ))
+            }
+        } else {
+            Err(Error::PositionOutOfBounds {
+                name: "attribute",
+                value: self.to_string(),
+                index,
+            })
+        }
+    }
+
+    /// Gets all attributes.
+    pub fn attributes(&'c self) -> impl Iterator<Item = (Identifier<'c>, Attribute<'c>)> {
```

Comment:
I don't know if it's correct to add lifetime `'c` to `self`, but it seemed necessary. Is this because the returned `Identifier` and `Attribute` live for `'c` and not for some `'_` that's shorter than `'c`, as is the case for the other getters?

Reply:
In my environment, I rather see a different error of opaque type lifetime for `impl Iterator` when I remove the `'c` lifetime from `&self`. Does this work on your side?

```suggestion
    pub fn attributes(&self) -> impl Iterator<Item = (Identifier<'c>, Attribute<'c>)> + '_ {
```

Reply:
That works as well! I didn't understand the error message well enough to fix it the right way.